sys/posix/semaphore.c

```diff
@@ -20,105 +20,45 @@
  * @}
  */
 
+#include <errno.h>
 #include <inttypes.h>
 
 #include "irq.h"
 #include "sched.h"
 #include "tcb.h"
+#include "timex.h"
 #include "thread.h"
+#include "vtimer.h"
 
 #define ENABLE_DEBUG (0)
 #include "debug.h"
 
 #include "semaphore.h"
 
-int sem_init(sem_t *sem, int pshared, unsigned int value)
-{
-    (void) pshared;  /* nothing to do */
-
-    sem->value = value;
-
-    /* waiters for the mutex */
-    sem->queue.first = NULL;
-
-    return 0;
-}
-
-int sem_destroy(sem_t *sem)
-{
-    if (sem->queue.first) {
-        DEBUG("sem_destroy: %" PRIkernel_pid ": tried to destroy active semaphore.\n",
-              sched_active_thread->pid);
-        return -1;
-    }
-    return 0;
-}
-
-sem_t *sem_open(const char *name, int oflag, ...)
-{
-    (void) name; /* named semaphore currently not supported */
-    (void) oflag;
-    return SEM_FAILED;
-}
-
-int sem_close(sem_t *sem)
-{
-    (void) sem; /* named semaphore currently not supported */
-    return -1;
-}
-
-int sem_unlink(const char *name)
-{
-    (void) name; /* named semaphore currently not supported */
-    return -1;
-}
-
-int sem_wait(sem_t *sem)
-{
-    while (1) {
-        unsigned old_state = disableIRQ();
-
-        unsigned value = sem->value;
-        if (value != 0) {
-            sem->value = value - 1;
-            restoreIRQ(old_state);
-            return 1;
-        }
-
-        /* I'm going blocked */
-        priority_queue_node_t n;
-        n.priority = (uint32_t) sched_active_thread->priority;
-        n.data = (uintptr_t) sched_active_thread;
-        n.next = NULL;
-        priority_queue_add(&sem->queue, &n);
-
-        DEBUG("sem_wait: %" PRIkernel_pid ": Adding node to mutex queue: prio: %" PRIu32 "\n",
-              sched_active_thread->pid, sched_active_thread->priority);
-
-        /* scheduler should schedule an other thread, that unlocks the
-         * mutex in the future, when this happens I get scheduled again
-         */
-        sched_set_status((tcb_t*) sched_active_thread, STATUS_MUTEX_BLOCKED);
-        restoreIRQ(old_state);
-        thread_yield_higher();
-    }
-}
+#define USEC_IN_NS  (100)
 
 int sem_timedwait(sem_t *sem, const struct timespec *abstime)
 {
-    (void) sem; /* timedwait currently not supported */
-    (void) abstime;
-    return -1; /* signal problem */
+    timex_t now, timeout = { abstime->tv_sec, abstime->tv_nsec / USEC_IN_NS };
+    vtimer_now(&now);
+    if (timex_cmp(now, timeout) > 0) {
+        return sem_wait(sem);
+    }
+    timeout = timex_sub(timeout, now);
+    return sem_wait_timed(sem, &timeout);
 }
 
 int sem_trywait(sem_t *sem)
 {
-    unsigned old_state = disableIRQ();
+    unsigned int old_state, value;
     int result;
-
-    unsigned value = sem->value;
+    if (sem == NULL) {
+        return -EINVAL;
+    }
+    old_state = disableIRQ();
```

Comment:
on second thought, it might not really give any benefits and only make the code less readable, ignore my comment for now. I did not think about the `if (value == 0)` case before.